src/main.rs

```diff
@@ -33,98 +34,112 @@ mod template;
+        eprintln!("{}", err);
+        std::process::exit(1);
+    }
+}
 
+fn try_main(args: &[&str]) -> Result<()> {
     match &args[1..] {
         ["--stdout"] | ["-s"] => generate_cargo_nix(io::stdout().lock()),
         ["--file"] | ["-f"] => write_to_file("Cargo.nix"),
         ["--file", file] | ["-f", file] => write_to_file(file),
         ["--help"] | ["-h"] => print_help(),
-        ["--version"] | ["-v"] => println!("{}", version()),
+        ["--version"] | ["-v"] => {
+            println!("{}", version());
+            Ok(())
+        }
         [] => print_help(),
         _ => {
             println!("Invalid arguments: {:?}", &args[1..]);
             println!("\nTry again, with help: \n");
-            print_help();
+            print_help()
         }
     }
 }
 
 fn version() -> Version {
-    Version::parse(env!("CARGO_PKG_VERSION")).expect("parse CARGO_PKG_VERSION")
+    // Since `CARGO_PKG_VERSION` is provided by Cargo itself, which uses the same `semver` crate to
+    // parse version strings, the `unwrap()` below should never fail.
+    Version::parse(env!("CARGO_PKG_VERSION")).unwrap()
 }
 
-fn read_version_attribute(path: &Path) -> Version {
-    let file = fs::File::open(path).expect(&format!("Couldn't open file {}", path.display()));
+fn read_version_attribute(path: &Path) -> Result<Version> {
+    let file = fs::File::open(path)
+        .with_context(|e| format!("Couldn't open file {}: {}", path.display(), e))?;
+
     io::BufReader::new(file)
         .lines()
         .filter_map(|line| line.ok())
         .find(|line| line.trim_start().starts_with(VERSION_ATTRIBUTE_NAME))
         .and_then(|s| {
             if let Some(i) = s.find('"') {
                 if let Some(j) = s.rfind('"') {
-                    return Some(Version::parse(&s[i + 1..j]).expect("parse version attribute"));
+                    return Version::parse(&s[i + 1..j]).ok();
                 }
             }
             None
         })
-        .expect(&format!(
-            "{} not found in {}",
-            VERSION_ATTRIBUTE_NAME,
-            path.display()
-        ))
+        .ok_or_else(|| {
+            failure::format_err!(
+                "valid {} not found in {}",
+                VERSION_ATTRIBUTE_NAME,
+                path.display()
+            )
+        })
 }
 
-fn version_req(path: &Path) -> (VersionReq, Version) {
-    let ver = read_version_attribute(path);
-    let requirement = format!(">={}.{}", ver.major, ver.minor);
-    (
-        VersionReq::parse(&requirement).expect(&format!(
-            "parse {} found in {}",
-            requirement,
-            path.display()
-        )),
-        ver,
-    )
+fn version_req(path: &Path) -> Result<(VersionReq, Version)> {
+    let version = read_version_attribute(path)?;
+    let req = format!(">={}.{}", version.major, version.minor);
+    VersionReq::parse(&req)
+        .map_err(|e| failure::format_err!("parse {} found in {}: {}", req, path.display(), e))
```

Comment:
Sure, I'll change that.

Reply:
IIRC `Error: From<Context<D>>` so we can use the `?` operator here to coerce the error type.

Reply:
Hmm, I'd like to avoid using `?` to keep the expression-like appearance of 104, but I just realized that we could do `map_err(Error::from)` after calling `context()`, so it's fine. 😄